File: data structure and algorithms/Graph Theory/3812.minimum-edge-toggles-on-a-tree.java
/*
 * @lc app=leetcode id=3812 lang=java
 *
 * [3812] Minimum Edge Toggles on a Tree
 *
 * need[u] says whether node u currently differs from target. Process leaves
 * upward; if child u still needs a flip, toggle the edge to its parent, which
 * fixes u and flips the parent's need. The root must end fixed.
 *
 * Java note: each adjacency entry stores neighbor and original edge index so
 * the sorted answer can return edge IDs.
 *
 * Time: O(n log n) because of final sorting. Space: O(n).
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

// @lc code=start
class Solution {
    public int[] minimumFlips(int n, int[][] edges, String start, String target) {
        List<int[]>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int i = 0; i < edges.length; i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            graph[u].add(new int[] {v, i});
            graph[v].add(new int[] {u, i});
        }

        int[] parent = new int[n];
        int[] parentEdge = new int[n];
        int[] order = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = -1;
            parentEdge[i] = -1;
        }
        int size = 0;
        order[size++] = 0;
        for (int i = 0; i < size; i++) {
            int node = order[i];
            for (int[] edge : graph[node]) {
                int nei = edge[0];
                if (nei == parent[node]) {
                    continue;
                }
                parent[nei] = node;
                parentEdge[nei] = edge[1];
                order[size++] = nei;
            }
        }

        int[] need = new int[n];
        for (int i = 0; i < n; i++) {
            need[i] = start.charAt(i) == target.charAt(i) ? 0 : 1;
        }

        List<Integer> answer = new ArrayList<>();
        for (int i = n - 1; i >= 1; i--) {
            int node = order[i];
            if (need[node] == 1) {
                answer.add(parentEdge[node]);
                need[node] ^= 1;
                need[parent[node]] ^= 1;
            }
        }
        if (need[0] == 1) {
            return new int[] {-1};
        }

        Collections.sort(answer);
        int[] out = new int[answer.size()];
        for (int i = 0; i < answer.size(); i++) {
            out[i] = answer.get(i);
        }
        return out;
    }
}
// @lc code=end

File: data structure and algorithms/Graph Theory/3812.minimum-edge-toggles-on-a-tree.go
package leetcode

//
// @lc app=leetcode id=3812 lang=golang
//
// [3812] Minimum Edge Toggles on a Tree
//
// Notes
// Interpret start/target mismatches as node parity needs. Root the tree and
// process leaves upward; if a child still needs a flip, toggle its parent edge,
// which fixes the child and flips the parent's need. The root must end even.
// Returned edge ids are sorted to match the Python implementation. Time: O(n log
// n) for sorting the answer. Space: O(n).
//
// @lc code=start

import "sort"

type edge3812 struct {
	to  int
	idx int
}

func MinimumFlips3812(n int, edges [][]int, start string, target string) []int {
	graph := make([][]edge3812, n)
	for idx, edge := range edges {
		u, v := edge[0], edge[1]
		graph[u] = append(graph[u], edge3812{to: v, idx: idx})
		graph[v] = append(graph[v], edge3812{to: u, idx: idx})
	}

	parent := make([]int, n)
	parentEdge := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = -1
		parentEdge[i] = -1
	}

	order := []int{0}
	for idx := 0; idx < len(order); idx++ {
		node := order[idx]
		for _, edge := range graph[node] {
			if edge.to == parent[node] {
				continue
			}
			parent[edge.to] = node
			parentEdge[edge.to] = edge.idx
			order = append(order, edge.to)
		}
	}

	need := make([]int, n)
	for i := 0; i < n; i++ {
		if start[i] != target[i] {
			need[i] = 1
		}
	}

	answer := []int{}
	for i := len(order) - 1; i >= 1; i-- {
		node := order[i]
		if need[node] == 1 {
			answer = append(answer, parentEdge[node])
			need[node] ^= 1
			need[parent[node]] ^= 1
		}
	}

	if need[0] == 1 {
		return []int{-1}
	}
	sort.Ints(answer)
	return answer
}

// @lc code=end
